### ISSUE TITLE
teams refuses to authenticate at all sometimes

### ISSUE THREAD

**Issue by liedekef** [IsmaelMartinez/teams-for-linux#740]
**Describe the bug**
teams refuses to authenticate at all sometimes. You can log in, but then it says something went wrong, you need to reauth, but when clicking on the login button again brings you back to the same screen.

**To Reproduce**
1. launch teams
2. try to log in
3. the message "something went wrong" appears with also "Error Code: interaction_required|AADSTS160021: Application requested a user session which does not exist."
4. In .xsession-errors I find this error:

```
Working directory "/home/xxx/[Invalid UTF-8]" does not exist. It won't be used when spawning "/opt/teams-for-linux/teams-for-linux".xdg-settings: default-url-scheme-handler not implemented for xfce
(node:107564) UnhandledPromiseRejectionWarning: ReferenceError: iconPath is not defined
    at applyAppConfiguration (/opt/teams-for-linux/resources/app.asar/app/mainAppWindow/index.js:87:44)
    at Object.onAppReady (/opt/teams-for-linux/resources/app.asar/app/mainAppWindow/index.js:59:2)
(Use `teams-for-linux --trace-warnings ...` to show where the warning was created)
(node:107564) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
```

**Desktop (please complete the following information):**
 - OS: fedora 37
 - Installation package rpm
 - Version teams-for-linux-1.0.60-1.x86_64

**Additional context**

**Comment by liedekef** [comment]
update: cleaning up the whole .config/teams-for-linux directory seems to fix it (but then I lose some settings of course ...)

**Comment by storm9c1** [comment]
I'm having the same issue, started happening today.

I wonder if it would let me proceed if it had the option to logout and start a new authentication session.  But that option is not available once stuck in the aforementioned "retry" loop.

OS: Ubuntu 22.04 LTS
Installed from snap store.
Version 1.0.55 (latest/stable)

Workaround: snap remove --purge teams-for-linux; snap install teams-for-linux

But as mentioned this workaround loses settings.  Especially device settings, which is hard to find these days (there is one under Calls).  But no longer available in the general Settings menu anymore.

**Comment by jijojosephk** [comment]
> I'm having the same issue, started happening today.
> 
> I wonder if it would let me proceed if it had the option to logout and start a new authentication session. But that option is not available once stuck in the aforementioned "retry" loop.
> 
> OS: Ubuntu 22.04 LTS Desktop: XFCE (X11) Installed from snap store. Version 1.0.55 (latest/stable)
> 
> Workaround: snap remove --purge teams-for-linux; snap install teams-for-linux
> 
> But as mentioned this workaround loses settings. Especially device settings, which is hard to find these days (there is one under Calls). But no longer available in the general Settings menu anymore.

You don't have to remove. You can launch it with `teams-for-linux --clearStorage` just once. This clears anything stored locally of-course settings too.

**Comment by jijojosephk** [comment]
@storm9c1 @liedekef what are the settings you'd like to keep even after clearing the storage? Authentication related issues we might not be able to fix as it's more controlled by MS. We can try to copy the settings over to the new session.

**Comment by jijojosephk** [comment]
I'll close this and open a new discussion on preserving settings.

**Comment by storm9c1** [comment]
Note that this issue still persists.  In the middle of a meeting today, I suddenly got logged out and then stuck in the login loop.

**Comment by storm9c1** [comment]
@jijojosephk Can we re-open this since the login problems still persist?

The only workaround is --clearStorage.  And today I got stuck in the login loop in the middle of a meeting.

I've upgraded to "1.0.68 rev 334 latest/stable" snap.

**Comment by jijojosephk** [comment]
Sorry that it happens. It's very difficult to find the pattern.

**Comment by storm9c1** [comment]
Hi @jijojosephk Is there anything I can do or provide that would help?  With the caveat being it's random and may be days before it happens again.

**Comment by jijojosephk** [comment]
> Hi @jijojosephk Can we re-open this since the login problems still persist?
> 
> The only workaround is --clearStorage.  And today I got stuck in the login loop in the middle of a meeting.
> 
> I've upgraded to "1.0.68 rev 334 latest/stable" snap.
> 
> Thanks!
> 
> 

Sure

**Comment by jijojosephk** [comment]
> Hi @jijojosephk Is there anything I can do or provide that would help?  With the caveat being it's random and may be days before it happens again.

You may not get anything from the usual app logs. When it happens, choose debug from tra icon and check the errors added to the console

**Comment by storm9c1** [comment]
@jijojosephk That would be a bit difficult when I don't have a tray icon.  I'm using XFCE and one never shows up.  Can I launch the app in debug mode?  Should I used the --webDebug option?  What level should I specify?  Thanks.

**Comment by storm9c1** [comment]
@jijojosephk Oh wait I can press "Alt" and get an Application menu there, which has a debug option available.  Should I use that instead?

**Comment by nneul** [comment]
@storm9c1 You should be able to add tray to any Xfce panel - that's what I have on mine:

![image](https://user-images.githubusercontent.com/828492/233201615-66bbd6e4-0ecf-4fed-bdf4-42401e5f9738.png)

It's "Notification Area" in Xfce.

**Comment by storm9c1** [comment]
@nneul Thanks!  I have "Notification Plugin", "Status Tray Plugin" and "Indicator Plugin" all installed and enabled.  I have other tray icons.  Just not Teams.  I also checked to make sure "Known indicators" are not set to hidden.

This is glitchy because I toggled the Application Indicators to unhidden and then back to hidden in the Indicator Plugin (to prevent duplicate icons).  It could also be that I had to restart with xfce4-panel -r as well.   Teams now appears in the Status Tray Plugin even with the Application Indicators hidden in the other plugin.  This is confusing.

For the record I like running multiple plugins because one of them, can't remember which, updates the Wifi icon with the signal strength and the other doesn't.  Not your concern.  Just explaining why.

However I guess it's plausible that it's causing a problem with Teams for Linux.

Let's not worry about that right now.  Since I now have two methods to enable debug mode at this point.

**Comment by storm9c1** [comment]
> > Hi @jijojosephk Can we re-open this since the login problems still persist?
> > The only workaround is --clearStorage.  And today I got stuck in the login loop in the middle of a meeting.
> > I've upgraded to "1.0.68 rev 334 latest/stable" snap.
> > Thanks!
> 
> Sure

I do not have GitHub permissions to reopen this.  I believe it would have to be re-opened by the submitter or by a contributor with proper authority.

**Comment by storm9c1** [comment]
Ah.  The problem just happened now!  I am stuck in the loop:

"Can you please sign in again?  Sorry for any inconvenience."

And the "Sign in" button flashes the window and puts me back in the same place.

I started it with --webDebug if that helps.  Where is the log file?

And if I press ALT and select "Debug" from the Application Menu, it starts the web console debugger/inspector.

I'll leave it stuck in this state today until I hear back from you.  Thanks!

**Comment by storm9c1** [comment]
![Screenshot_2023-04-20_14-51-31](https://user-images.githubusercontent.com/16827398/233460671-53b7dc7e-26f7-4b55-ac32-d0bcf58c4871.png)

**Comment by nneul** [comment]
Not sure if this helps discussion any - but - I'm getting the "lost session" behavior just about daily now - but I am _not_ getting the "stuck and can't re-authenticate". I _DID_ get that stuck behavior once on both of my client deployments, but hasn't seen it since. 

@jijojosephk Any chance this is related to changes recently on the 'active' check - did that have any behavioral change that would have cause the client to "stop talking to MS as regularly as it did before" resulting in session going stale or something like that?

**Comment by storm9c1** [comment]
[errors.txt](https://github.com/IsmaelMartinez/teams-for-linux/files/11289563/errors.txt)

**Comment by jijojosephk** [comment]
Sorry, I'm catching a flight now. I'll have a look at the log. From the screenshot, I suspect the app is worried that the presence state is updated without any user interaction.

**Comment by jijojosephk** [comment]
Wondering why I'm not getting the same issue.

**Comment by storm9c1** [comment]
> Wondering why I'm not getting the same issue.

Do you stay connected and logged in for days at a  time?  I do.

Could also be the organization asking for re-auth after some time passes.  And that breaks something.  I know the old thick client asked me to re-auth every once in a while.  As often as every 24 hours... or as infrequently as a week.  I never could find a pattern there.

**Comment by nneul** [comment]
I'm Xfce on Ub20 - and am seeing across two organizations. One that uses OneLogin for auth, one that is pure AzureAD. I see the symptom with both orgs, across two systems (work and home) - that both stay logged in for extended periods.

**Comment by svili** [comment]
Have the exact same problem, started happening after password change to organization account.

**Comment by joostoudeman** [comment]
I just had the same thing happen to me.
I temporarily used a browser session after restarting the app several times and the problem persisting..
After the meeting I started the app again, and, the problem apparently fixed itself. Not sure if this information is helpful, but I thought I should mention it.

**Comment by jijojosephk** [comment]
@liedekef this following error seems like a glitch in the code. Because you're using `--enableDesktopNotificationsHack` I believe. I'll remove the use of the flag because it's no more required. We'll see if the problem still persists after remove that also.
`(node:107564) UnhandledPromiseRejectionWarning: ReferenceError: iconPath is not defined
    at applyAppConfiguration (/opt/teams-for-linux/resources/app.asar/app/mainAppWindow/index.js:87:44)
    at Object.onAppReady (/opt/teams-for-linux/resources/app.asar/app/mainAppWindow/index.js:59:2)`

**Comment by jijojosephk** [comment]
@liedekef can you check in `1.0.77` ?

**Comment by storm9c1** [comment]
Hi @jijojosephk , FWIW, I am not using `--enableDesktopNotificationsHack`.

**Comment by liedekef** [comment]
I was using the --enableDesktopNotificationsHack option, but the reauth loop only happens once in a while, so difficult to test/check :-)

**Comment by jijojosephk** [comment]
@liedekef Thanks, `enableDesktopNotificationsHack` is no more required. Even if you pass it won't be considered. So the error should not be there in the new version. The re-auth loop I had also faced but it's been a while. I'm pretty sure it has something to do with organisational security settings.

**Comment by storm9c1** [comment]
I recommended teams-for-linux to a colleague that works at my same organization.  And of course today he ran into the same auth-loop issue.

He also mentioned the same thing about -clearStorage not preserving dark mode and device settings.  Unsure if there is a way when using a sledge hammer like that.

But maybe we can add an option to just clear the auth cookies or flush whatever might be causing the auth problems without nuking all of the cache storage?

**Comment by jijojosephk** [comment]
When I had faced the same in the past I tried the developer tool and cleared some cookies still it continued to have the loop. It's the same behaviour in google chrome as well. Thinking out loud.

**Comment by jijojosephk** [comment]
Can you confirm running this on developer tools console when this issue happens helps you log out? It might show a dialogue asking are you sure to logout.

```js
const controller = typeof window.angular !== 'undefined' ? window.angular.element(document.documentElement).controller() : null;
controller.authenticationService.logOut();
```

**Comment by storm9c1** [comment]
Sure, will try that next time it happens.  Honestly it hasn't happened since upgrading to 1.0.80 so far but it's only been a few days.

Thanks for the quick response.

**Comment by jijojosephk** [comment]
> I recommended teams-for-linux to a colleague that works at my same organization. And of course today he ran into the same auth-loop issue. He is using RHEL 9. And Fedora 37.
> 
> He also mentioned the same thing about -clearStorage not preserving dark mode, geometry, and device settings. Unsure if there is a way when using a sledge hammer like that.
> 
> But maybe we can add an option to just clear the auth cookies or flush whatever might be causing the auth problems without nuking all of the config cache storage?

What comes in geometry?

**Comment by storm9c1** [comment]
Window position and size when Teams starts up.

**Comment by jijojosephk** [comment]
@storm9c1 `--clearStorage` does not remove window size, position, it preserves zoom level also. Please report if you have a different experience.

**Comment by storm9c1** [comment]
You know what, I haven't seen that behavior since I last used "snap remove --purge teams-for-linux".

Once I started using --clearStorage, it seems that the geometry is mostly preserved.  I say mostly because it's a bit off in the vertical positioning.  It positions about 1cm lower on my monitor, about the width of the title bar.  But that happens even if I just close and re-open it.  So that seems unrelated.

My colleague told me that he killed it and used "rm -rf ~/.config/teams-for-linux" to get out of the login loop.  I suggested to use --clearstorage in the future.  :-)

**Comment by jijojosephk** [comment]
Thanks, I'll work on how to get some important settings preserved. Very difficult to catch at what point MS sets the theme, device settings when the app loads. There can be multiple ways.

**To save**

1. Read the local, session storage db. Not sure possible or
2. Wait for the app to load and execute method to read settings and store locally. Like other hacks work.

**To restore**

1. Write locally stored settings to apps db. Again don't know how it works.
2. Wait for the app to load and execute method to set values from local store.

May be while app runing.

- `teams-for-linux --saveSettings`
- Quit the app.
- `teams-for-linux --clearStorage`
- After the app loaded.
- `teams-for-linux --restoreSettings`

**Comment by IsmaelMartinez** [comment]
We can probably just copy the config file, clear the cache and then restore
the config file and restart the app?

**Comment by jijojosephk** [comment]
`1.0.83` is out. It has more menu options. Check the release page.

@liedekef and @storm9c1 when you face the loop, what can be done now is:

- Choose Settings->Save
- Choose Quit->Clear Storage
- Restart the app and complete login
- Choose Settings->Restore

**Comment by storm9c1** [comment]
@jijojosephk Thank you!  Great work!  I will try it next time I hit the issue.

I hit it today in the middle of a meeting, which means I was in a hurry to get back on and couldn't do any further debugging.  Sorry.  It's really a panic when it happens in a meeting.

Would it be a valuable incremental improvement to make "Save" and "Restore" command line switches so that I can automate it from a script?

I'm thinking something like this:
teams-for-linux --settingsBackup --noLaunch
teams-for-linux --clearStorage --noLaunch
teams-for-linux --settingsRestore --noLaunch

Whereas --noLaunch won't actually fire up the app, it will just do the storage maintenance.

**Comment by jijojosephk** [comment]
My initial attempt was to implement through command line but without launching the getting the settings is pretty difficult. Very difficult to know where these settings are stored. Secondly, after clearing the storage, the default values are coming from MS server, so difficult to catch when these settings are applied. Command line looks nearly impossible.

**Comment by storm9c1** [comment]
Ah OK, well it doesn't hurt to ask.  Thanks anyways.

**Comment by storm9c1** [comment]
Note that I haven't forgotten to run that code snippet (above) from the developer console next time this happens.  I just couldn't today because I needed to get back into the meeting immediately.

**Comment by jijojosephk** [comment]
The root cause for auto loop is still unknown. Can you check if that happens after resuming from standby?

**Comment by storm9c1** [comment]
> The root cause for auto loop is still unknown. Can you check if that happens after resuming from standby?

This is on a desktop machine, not a laptop.  I have no power saving modes enabled (nor do I want to).  I also tried disabling screen locking for a while to see if that made a difference.  Nope.

**Comment by jijojosephk** [comment]
After very long time, I got this yesterday but didn't face the looping. Clicked on on sign-in and it was ok.

**Comment by jijojosephk** [comment]
> > The root cause for auto loop is still unknown. Can you check if that happens after resuming from standby?
> 
> This is on a desktop machine, not a laptop. I have no power saving modes enabled (nor do I want to). I also tried disabling screen locking for a while to see if that made a difference. Nope.
> 
> Update: I also have "--appIdleTimeout 86400". But I'm not sure that's working right.

It definitely delays user presence turning orange. But the session timeout is something different. What I believe is some user activity is needed like at least the app getting focus after certain time, clicking somewhere. If we're able to the simulate the same or find where's that flag and reset, we'd be successful. But it requires quite a lot of time.

**Comment by storm9c1** [comment]
>> Update: I also have "--appIdleTimeout 86400". But I'm not sure that's working right.

> It definitely delays user presence turning orange. But the session timeout is something different.

I figured they weren't necessarily related.  But I mentioned it because if I `cat /proc/PID/cmdline`, or check "ps" output, it shows `--appIdleTimeout` but not the integer argument.

If I use `--appIdleTimeout=XXX` (adding the "=" char) then it shows up in the /proc/PID/cmdline and in "ps".  But I'm not sure which version the command line parser expects.

Also if I move focus away for a while, Teams doesn't automatically refresh after a while.  As soon as I give it focus again, it suddenly catches up.  I wasn't sure if the `--appIdleTimeout` is supposed to affect that behavior or not.

If you think this may be a separate defect, I'm happy to file a new bug.  Otherwise we can ignore this.

Maybe a future improvement is to just send fake mouse events to Teams until the timeout is reached.  I actually did this in a crude way with the Microsoft Teams thick client, using xdotool to move the mouse a pixel back and fourth, to keep it alive so Teams (and my machine) wouldn't go idle during multi-hour meetings or chat sessions.  But that was a different scenario.

PS: I haven't had the original login loop issue all week now.  :-)

Thanks again for following up and your hard work.  Much appreciated.

**Comment by jijojosephk** [comment]
Thanks for the update. Can we close this?

**Comment by liedekef** [comment]
For me this can be closed.

**Comment by storm9c1** [comment]
Yes, I'm comfortable with closing this at this point.  Thanks!

**Comment by jijojosephk** [comment]
Thanks @liedekef and @storm9c1 for testing this. I'm closing this now. Could you please test these mysterious ones too ? :smile: #699 #724 The latter being absolutely weird.

**Comment by storm9c1** [comment]
> Thanks @liedekef and @storm9c1 for testing this. I'm closing this now. Could you please test these mysterious ones too ? smile #699 #683 #724 The latter being absolutely weird.

Sure, I will weigh-in on those other issues in their respective threads.

**Comment by storm9c1** [comment]
And of course, today I get stuck in the login loop.  (facepalm)

No need to re-open this.  I give up.

I just followed the steps mentioned in [this comment](https://github.com/IsmaelMartinez/teams-for-linux/issues/740#issuecomment-1534115693).

And all was well after a full reauth cycle.

**Comment by storm49152** [comment]
I know this is closed, but for the record: I'm on 1.1.16, and I am experiencing this issue at least a couple of times per week. I'm seeing the same errors in the debug log as in the screenshot. The save, quit/clear, restore sequence works, so that's good and I'm not requesting to re-open.

Some background:

The organization I'm currently working for is using user/pass login together with the MS authenticator, and requires us to re-login every 3 or so hours. The most annoying thing about this, is that when you're not watching the Teams window, you'll never know until you need Teams. But yeah, we're "encouraged" to use it anyway, because of "reasons". Oh, well...

When this happens, teams-for-linux presents a popup asking you to login again. When I enter the password (sporadically I don't have to, and it "just works" after clicking...?), the MS authenticator on the phone should alert and ask for approval of the login. (This does not always work, as sometimes the MS authenticator doesn't show an alert, but that's not teams-for-linux's fault.) After the login is approved, it works most of the time. Other times I'll enter the login loop. When that happens, the application changes window content a couple of times, and in between it says something about "changing the browser". But that message goes away too fast to make a screenshot or so to show the exact message.

**Comment by storm9c1** [comment]
Wow, 3 hours is rough.  I think ours is 24 hours which is inconvenient but at least I can get through a work day.

I'd say 1 out 20 times I'd get stuck in the login loop still.

Which is about once every two weeks or so.  And the "clear storage" workaround fixes it.

Unfortunately, no matter what Teams client is used, if your organization sets ridiculous policies, "we can't fix stupid."

Maybe a future improvement would be either a desktop notification and a special sound to indicate that you've been thrown out.  I know this won't fix the auth loop but maybe it will help the overall experience.  Well, maybe there's a small chance it will fix the auth loop since perhaps I can re-auth immediately when I am notified.  Rather than waiting several hours and by that time, the auth may expire.

Food for thought.
